src/Simulation/Common/QubitManager.cs

```diff
@@ -74,185 +166,224 @@ public class QubitManager
-                }
+                operationStack = new Stack<StackFrame>();
+                curFrame = new StackFrame();
             }
+        }
 
-            for (long i = oldNumQubits; i < NumQubits; i++)
-            {
-                this.qubits[i] = i + 1;
-            }
-            Debug.Assert(this.qubits[NumQubits - 1] == None);
+        private class QubitNonAbstract : Qubit
+        {
+            // This class is only needed because Qubit is abstract. 
+            // It is equivalent to Qubit and adds nothing to it except the ability to create it.
+            // It should be used only in CreateQubitObject below, and nowhere else.
+            // When Qubit stops being abstract, this class should be removed.
+            public QubitNonAbstract(int id) : base(id) { }
+        }
 
-            if (free == oldNone)
+        /// <summary>
+        /// May be overriden to create a custom Qubit object of a derived type.
+        /// </summary>
+        /// <param name="id">unique qubit id</param>
+        /// <returns>a newly instantiated qubit</returns>
+        public virtual Qubit CreateQubitObject(long id)
+        { 
+            if (id >= Int32.MaxValue)
             {
-                free = oldNumQubits;
-                freeTail = NumQubits - 1;
+                throw new NotSupportedException($"Qubit id out of range.");
             }
+            return new QubitNonAbstract((int)id);
         }
 
-        public bool IsValid(Qubit qubit)
+        public virtual bool IsValid(Qubit qubit)
         {
             if (qubit == null) { return false; }
             if (qubit.Id >= NumQubits) { return false; }
             if (qubit.Id < 0) { return false; }
             return true;
         }
 
-        public bool IsFree(Qubit qubit)
-        {
-            if (!IsValid(qubit)) { return false; }
-            return IsFree(qubit.Id);
-        }
-
         private bool IsFree(long id)
         {
             return qubits[id] < Allocated;
         }
 
-        public bool IsDisabled(Qubit qubit)
+        public virtual bool IsFree(Qubit qubit)
         {
-            if (!IsValid(qubit)) { return false; }
-            return IsDisabled(qubit.Id);
+            return IsValid(qubit) && IsFree(qubit.Id);
         }
 
         private bool IsDisabled(long id)
         {
             return (qubits[id] == Disabled);
         }
 
+        public virtual bool IsDisabled(Qubit qubit)
+        {
+            return IsValid(qubit) && IsDisabled(qubit.Id);
+        }
+
         private bool IsAllocatedForBorrowing(long id)
         {
             return qubits[id] >= AllocatedForBorrowing;
         }
 
-        // Returns true if qubit needs to be released: It has been allocated for borrowing and is not borrowed any more.
-        private bool BorrowingRefCountIsOne(long id)
+        public virtual long GetQubitsAvailableToBorrowCount(int stackFrame)
```

Comment:
Since I don't think this method is called very often or with a large number here, I am not too concerned about the overhead of the temporary stack. Having the option to query the available qubits to borrow for the current frame makes a lot of sense to me, and I could imaging being able to query other frames would be useful - that is if we are not talking about the scenario where these are invoked from within Q# but rather directly by something interacting with the manager.

Reply:
OK. It's not so much a perf issue as a code complexity issue, but it's not particularly complicated as it is :-).

Reply:
It got even simpler with having the current frame be part of the stack.